File: examples/auction/become.rb
#!/usr/bin/env ruby

# cf. Scala by Example, Chapter 3

$:.push File.join( File.dirname(__FILE__), "..", "..", "lib" )

require 'dramatis/actor'
require 'dramatis/actor/behavior'

module Auction

  def self.new *args
    Open.new( *args )
  end

  class Open
    include Dramatis::Actor
    def initialize seller, min_bid, closing
      @seller = seller
      @min_bid = min_bid
      @closing = closing

      @bid_increment = 10
      @max_bid = @min_bid - @bid_increment
      @max_bidder = nil

      actor.refuse :winner

      release( actor.name ).close
    end

    def close
      actor.yield @closing - Time::now

      if @max_bid > @min_bid
        release( @seller ).winner @max_bidder
        release( @max_bidder ).winner @seller
        actor.become Over.new( @max_bidder, @max_bid )
      else
        release( @seller ).failed @max_bid
        actor.become Over.new( nil, @max_bid )
      end
    end

    def inquire
      [ @max_bid, @closing ]
    end

    def offer bid, bidder
      if bid >= @max_bid + @bid_increment
        if @max_bid >= @min_bid
          release( @max_bidder ).beaten_offer bid
        end
        @max_bid = bid
        @max_bidder = bidder
        :best_offer
      else
        [ :beaten_offer, @max_bid ]
      end
    end
  end

  class Over
    include Dramatis::Actor::Behavior
    attr_reader :winner, :max_bid
    def initialize winner, max_bid
      @winner = winner
      @max_bid = max_bid
    end
    def dramatis_bound
      actor.accept :winner
    end
    def inquire *args
      [ @max_bid, 0 ]
    end
    def offer *args
      :auction_over
    end
  end

end

class Seller
  include Dramatis::Actor
  def winner winner
  end
  def failed highest_bid
  end
end

class Client
  include Dramatis::Actor
  attr_reader :name
  def initialize name, increment, top, auction
    @name = name
    @increment = increment
    @top = top
    @auction = auction
    @current = 0
    log "started"
    @max = auction.inquire[0]
    log "status #{@max}"
    bid
  end
  def bid
    if @max > @top
      log("too high for me")
    elsif ( @current < @max )
      @current = @max + @increment
      sleep( ( 1 + rand( 1000 ) )/1000.0 )
      answer, max_bid = @auction.offer @current, actor.name
      case answer
      when :best_offer; log("best offer: #{@current}")
      when :beaten_offer; beaten_offer max_bid
      when :auction_over; log("auction over, oh well")
      end
    end
  end
  def beaten_offer max_bid
    log("beaten offer: #{max_bid}")
    @max = max_bid
    release( actor.name ).bid
  end
  def winner seller
    log "I won!"
  end
  def log string
    puts "client #{@name}: #{string}"
  end
end

# somebody gives up

seller = Seller.new
auction = Auction.new seller, 100, Time::now + 4
Client.new "1a", 20, 200, auction
Client.new "2a", 10, 300, auction

puts "Notice: client #{auction.winner.name} won the first auction with a bid of #{auction.max_bid}"

# cut off while people still have money

seller = Seller.new
auction = Auction.new seller, 100, Time::now + 1.5
Client.new "1b", 20, 200, auction
Client.new "2b", 10, 300, auction

puts "Notice: client #{auction.winner.name} won the second auction with a bid of #{auction.max_bid}"

# too expensive

seller = Seller.new
auction = Auction.new seller, 400, Time::now + 1.5
Client.new "1c", 20, 200, auction
Client.new "2c", 10, 300, auction

raise RuntimeError if auction.winner != nil

puts "Notice: the third auction failed; the maximum recieved bid was #{auction.max_bid}"

File: examples/auction/become.py
#!/usr/bin/env python

# cf. Scala by Example, Chapter 3

import time
import random
import inspect
import sys
import os.path

sys.path[0:0] = [ os.path.join( os.path.dirname( inspect.getabsfile( inspect.currentframe() ) ), '..', 'lib' ) ]

import dramatis

class Auction(object):

    def __new__( *args ):
        return Object( *args )

    class Open( dramatis.Actor ):

        def __init__(self, seller, min_bid, closing):
            self._seller = seller
            self._min_bid = min_bid
            self._closing = closing
            
            self._bid_increment = 10
            self._max_bid = self._min_bid - self._bid_increment
            self._max_bidder = None

            self.actor.refuse "winner"
            
            dramatis.release( self.actor.name ).close()

        def close(self):
            self.actor.yield( @closing - Time::now )

            if self._max_bid > self._min_bid:
                dramatis.release( self._seller ).winner( self._max_bidder )
                dramatis.release( self._max_bidder ).winner( self._seller )
                self.actor.become Over( self._max_bidder, self._max_bid )
            else:
                dramatis.release( self._seller ).failed( self._max_bid )
                self.actor.become Over.new( nil, self._max_bid )
                
                def inquire(self):
                    return [ self._max_bid, self._closing ]

        def offer(self, bid, bidder):
            if bid >= self._max_bid + self._bid_increment:
                if self._max_bid >= self._min_bid:
                    dramatis.release( self._max_bidder ).beaten_offer( bid )
                self._max_bid = bid
                self._max_bidder = bidder
                return [ "best_offer", None ]
            else:
                return [ "beaten_offer", self._max_bid ]

    class Over( dramatis.Actor.Behavior ):
        @propery
        def winner(self): return self._winner
        @propery
        def max_bid(self): return self._max_bid

        def __init__(self, winner, max_bid):
            self._winner = winner
            self._max_bid = max_bid

        def dramatis_bound(self):
            self.actor.accept( "winner" )

        def inquire(self, *args):
            [ self._max_bid, 0 ]

        def offer(self, *args):
            return "auction_over"

class Seller ( dramatis.Actor ):
    def winner(self, winner): pass
    def failed(self, highest_bid): pass

class Client ( dramatis.Actor ):
    @property
    def name(self): return self._name

    def __init__(self, name, increment, top, auction):
        self._name = name
        self._increment = increment
        self._top = top
        self._auction = auction
        self._current = 0
        self.log( "started" )
        self._max = auction.inquire()[0]
        self.log( "status " + str(self._max) )
        self.bid()

    def bid(self):
        if self._max > self._top:
            self.log("too high for me")
        elif ( self._current < self._max ):
            self._current = self._max + self._increment
            time.sleep( ( 1 + random.randint( 0, 1000 ) )/1000.0 )
            answer, max_bid = self._auction.offer( self._current,
                                                   self.actor.name )
            if answer == "best_offer":
                self.log("best offer: " + str(self._current))
            elif answer == "beaten_offer": self.beaten_offer( max_bid )
            elif answer == "auction_over":
                self.log("auction over, oh well")

    def beaten_offer(self, max_bid):
        self.log("beaten offer: " + str(max_bid))
        self._max = max_bid
        dramatis.release( self.actor.name ).bid()

    def winner(self, seller):
        self.log("I won!")

    def log(self,string):
        print "client", self._name, ": ", string

# Long enough to resolve

seller = Seller()
auction = Auction( seller, 100, time.time() + 4 )
Client( "1a", 20, 200, auction )
Client( "1b", 10, 300, auction )

puts "Notice: client", auction.winner.name, "won the first auction with a bid of", auction.max_bid
